### ISSUE TITLE
Stage constructor seems to leave `dim_vars` uninitialized.

### ISSUE THREAD

**Issue by alexreinking** [halide/Halide#4030]
In the string-args constructor for Stage, a local variable named `dim_vars` is created that shadows the class field.

https://github.com/halide/Halide/blob/fa25a98119c9a8dc4a3b4f758569966c5bd50b35/src/Func.h#L95-L99

That sequence should be replaced with a `reserve` and `emplace_back` loop a la:

```cpp
dim_vars.reserve(args.size());
for (const auto &arg : args) {
    dim_vars.emplace_back(arg);
}
```

If this looks right to everyone else, I'll open a PR. Regardless, the code sequence is a no-op unless the assertion fails. If we want the assertion, the loop could be dropped while maintaining the existing behavior.

When might this cause a problem, I wonder?

**Comment by abadams** [comment]
Yikes. Code that uses that constructor doesn't seem like it could work. Maybe try just deleting that constructor first?

**Comment by alexreinking** [comment]
The only usages are in FuncRef's operator=, here:

https://github.com/halide/Halide/blob/fa25a98119c9a8dc4a3b4f758569966c5bd50b35/src/Func.cpp#L2641-L2648

These could be modified to create Var vectors instead of string vectors and use the first constructor.

**Comment by alexreinking** [comment]
Actually, I'm 99% convinced that fourth argument is unnecessary and it can always be derived from the `Function`'s `args()` method. I'm going to try that change locally.

**Comment by alexreinking** [comment]
Yup, l I was right. Updated the PR. PTAL :slightly_smiling_face:

**Comment by steven-johnson** [comment]
Is this fixed?

**Comment by alexreinking** [comment]
@steven-johnson Yes. Fixed in #4031
